### ISSUE TITLE
Global query: wrong heads extraction

### ISSUE THREAD

**Issue by shkarupa-alex** [NVlabs/GCVit#11]
Model code has an issue in adapting global query for attention estimation: https://github.com/NVlabs/GCVit/blob/main/models/gc_vit.py#L223
Heads in keys and values and local query produced by splitting channels and head dimension transposition to "batch" dimension (in terms of matmul).
Heads in global query produced by shifting spatial size https://github.com/NVlabs/GCVit/blob/main/models/gc_vit.py#L224 because of missing transpose op that moves head dimension.

As far as i understand this results in not full query-to-key spatial interactions and adds some query-key channel shifting.

**Comment by ahatamiz** [comment]
@shkarupa-alex thanks for the comment. In this case, heass for global query is still generated by splitting across channel and head ( see the global query token generator regarding the output shape).

**Comment by shkarupa-alex** [comment]
> heass for global query is still generated by splitting across channel and head
You are definitely wrong.
1) q_global here https://github.com/NVlabs/GCVit/blob/main/models/gc_vit.py#L224 has shape **[batch, channel, height, width]** (height and width both equals to window size = 7)
2) Repeating across batch dimension in the next line does not change shape mening
3) Then you reshape q_global in [batch, num_heads, height * width, channel // num_heads]

Let's take an example:
batch = 2
channel = 30
height = width = window_size = 7
num_heads = 3

So you take [2, 30, 7, 7] tensor, reshape it to [2, 3, 7 * 7, 10] and think that last 2 dimensions are length and channels.
But it is wrong. You mix values from different channels in different height/width splits.

The right way is: 
[2, 30, 7, 7] -> reshape to extract heads [2, 3, 10, 7, 7] -> transpose to put channels at last axis [2, 3, 7, 7, 10] -> reshape to flatten height and width as length [2, 3, 7*7, 10]

**Comment by rwightman** [comment]
@ahatamiz as per @shkarupa-alex I'm also don't see how this works. 

Regardless of how you look at #13 and the fact that the global context is completely jumbled by reinterpreting the NHWC tensor as NCHW when it passes to the FeatureExtract stack, g_global being NCHW OR NHWC (given #13, what is it supposed to be?) doesn't map cleanly to the reshape used, channels and spatial dims are being mixed up in either case...

**Comment by ahatamiz** [comment]
Hi @shkarupa-alex and @rwightman , thank you so much for the insightful comments which will allow us to make GC ViT even better. We will investigate and update all on the status.

Best,

**Comment by pamolchanov** [comment]
We indeed refactored our internal version of the code and introduced some bugs trying to simplify things. Thanks @rwightman for pointing to the bug in the incorrect reshaping the input to FeatureExtract. We have fixed it to be in NCHW for FeatureExtract, while the rest of transformer blocks to be in NHWC with proper dimension permutation. The correct version of the code runs fine and we are retraining models to update weight together with the code. Expectation is to finish by EOW.

**Comment by ahatamiz** [comment]
We have updated our models and re-trained them. In all cases, the accuracy is better compared to reported numbers in the paper, from our internal checkpoints. In this version, we have converted tensors to be in `nhwc` dimensions, and only be `bchw` for conv layers. In addition, we have conducted a set of stress tests and verified there is no inconsistency with the model performance across different batch sizes at test time. The new version of the model shows improvement of 0.15 on average over the previous model across all 5 variants. We would like to thank @rwightman , @shkarupa-alex and @awsaf49  for pointing out the issues with the initial release.

Best,

**Comment by leondgarse** [comment]
Isn't it still reshaping a channel first global query as if it's channel last in [gc_vit.py#L515-L518](https://github.com/NVlabs/GCVit/blob/main/models/gc_vit.py#L515-L518)?
```py
def forward(self, x):
    x = self.to_q_global(x)
    B = x.shape[0]
    x = x.reshape(B, 1, self.N, self.num_heads, self.dim_head).permute(0, 1, 3, 2, 4)
```
In my understanding, we firstly have input x shape `[batch, channels, window_size, window_size]`, then reshaped to `[batch, 1, window_size * window_size, num_heads, channels // num_heads]`. Is it intend to like this or I'm missing something here?

**Comment by awsaf49** [comment]
https://github.com/NVlabs/GCVit/blob/ea880cb0e5e5493096c0ecb87eb59998d229122f/models/gc_vit.py#L516-L518
the output of `to_q_global`, `x` is `channel_first` due to `nn.Conv2d` but here `reshape` expects output to be `channel_last` so perhaps @leondgarse  is suggesting something like this?
```py
 def forward(self, x):
    x = self.to_q_global(x) # (B, C, H, W)
    x = _to_channel_last(x)  # (B, H, W, C)
    B = x.shape[0]
    x = x.reshape(B, 1, self.N, self.num_heads, self.dim_head).permute(0, 1, 3, 2, 4)  # (B, 1, num_heads, N, dim_head)
```

**Comment by shkarupa-alex** [comment]
+1 to "issue in channel/spatial mixing"
**input x** is [batch, channel, height, width]
**x after to_q_global** is [batch, channel, window, window]

As @awsaf49 mentioned there should be 
either conversion to channel first before reshaping
or `x = x.reshape(B, 1, self.num_heads, self.dim_head, self.N).permute(0, 1, 2, 4, 3)`

**Comment by ahatamiz** [comment]
Hi @leondgarse, @awsaf49  and @shkarupa-alex 

Thank your for your comments. With or without conversion to channel first before reshaping, the output of both operations are equivalent. As a sanity check. I have written the following snippet which you can verify. `q_global` is a randomly initialized channel last tensor while `q_global2` is the same reshaped to channel first.

```
import torch

q_global = torch.rand(size=(1, 7, 7, 64)).cuda()
q_global_copy = q_global.clone()  # initial cloned q_global [B,H,W,C]
q_global2 = q_global_copy.permute(0, 3, 1, 2)  # reshaped cloned q_global [B,C,H,W]

B = 1
N = 49
num_heads = 2
C = 64
B_ = 64
q_global = q_global.reshape(B, 1, N, num_heads, C // num_heads).permute(0, 1, 3, 2, 4)
q_out = q_global.repeat(1, B_ // B, 1, 1, 1).reshape(B_, num_heads, N, C // num_heads)


q_global2 = q_global2.reshape(B,  num_heads, C // num_heads, -1).permute(0, 1, 3, 2)
q_out2 = q_global2.repeat_interleave(B_//B, dim=0)

print(torch.equal(q_out, q_out2))
```

Hence, there feature extraction in global query generator is done correctly, and provided checkpoints reflect the intended implementation. 

We encourage your to try the newly provided checkpoints. As stated earlier, the performance have improved across all models, sometimes by a significant margin, hence setting new SOTA performance in terms of Image-Net Top1 accuracy. 

Best,

**Comment by leondgarse** [comment]
Besides, I think the repeating method still not changed:
```py
aa = np.arange(6).reshape(2, 3)
print(aa)
# [[0, 1, 2],
#  [3, 4, 5]]

""" Previous repeat method """
print(torch.from_numpy(aa).repeat(2, 1))
# tensor([[0, 1, 2],
#         [3, 4, 5],
#         [0, 1, 2],
#         [3, 4, 5]])

""" Current method """
print(torch.from_numpy(aa[:, None, :]).repeat(1, 2, 1).reshape(4, 3))
# tensor([[0, 1, 2],
#         [3, 4, 5],
#         [0, 1, 2],
#         [3, 4, 5]])
```

**Comment by pamolchanov** [comment]
@leondgarse @shkarupa-alex you are right. I can confirm we had to do channel_last before reshape in GlobalQueryGen. We had it before there, but somehow it got disappeared. We are fixing it now. However, it is quite interesting that reshaping unit tests are fine, and I think there is something in there suggesting that noisy attention is beneficial for transformers. Thanks for spotting it, really appreciate.

**Comment by leondgarse** [comment]
Ya, even the previous mixing batches works, and it's almost a random query... I think it's a clue that maybe the attention blocks can be further simplified here.

**Comment by ahatamiz** [comment]
@leondgarse Our internal checkpoints remain intact and we have verified the effectiveness o the global query tokens. We will provide new models with the refactored version very soon. 

Please stay tuned for the new models.

**Comment by awsaf49** [comment]
@ahatamiz  So, the recently updated score on readme reflects to the intact checkpoints?

**Comment by ahatamiz** [comment]
@awsaf49 Yes. We will provide the statistics and checkpoints for the refactored version very soon. 

Thanks.

**Comment by leondgarse** [comment]
Here's my testing result using current shared checkpoints:
| Model | Top Acc (current code base reshape method) | Top 1 Acc (expecting refactored method) |
|-|-|-|
| GC ViT-T | top1: 83.46 top5: 96.48 | top1: 82.242 top5: 95.154  |
| GC ViT-S | top1: 83.952 top5: 96.68 | top1: 82.662 top5: 96.06 |

So ya, we can look forward for your new checkpoints. :)

**Comment by ahatamiz** [comment]
@leondgarse the left column is our own reported numbers. The right column sound mystical. I recommend that you wait for the announcement regarding the weights. 

Thanks

**Comment by rwightman** [comment]
I guess it just goes to show how liberal the definition of 'mixing' can be, and how suspect many justifications of exotic attention/mixing mechanisms are. It's not that hard to get this right, I have a local gcvit impl with the correct tensor layouts. 

The numbers are decent but I wouldn't say gcvit breaks away from the pack. It's slow like swin, easier to train like swin, sure, but well trained vanilla vit can still manage comparable accuracy with lower runtime cost. While the param counts and flops for gcvit are decent, like swin it has very high activation counts (as a result mem bandwidth limited) compared to vanilla vit.

**Comment by pamolchanov** [comment]
Those are fair observations. We followed the next line of thoughts that potentially others follow as well. We want to study and advance transformers. At the first look we see that the complexity of vit scales quadratically with image size. One of the solutions to that is swin as it operates on local windows. Cross window information fusion happens with shifts. As a result, global information is not used as shifts don't allow for full interaction at every step. Therefore, we thought  of creating a global attention module that will enable that. Finally, cnns rely on inductive bias of structured data and we want to use it in the transformer, and add it to the global query generator. 
 
I see the following learning from 2 releases. If q is random (verified by perturbing the batch and all elements) the attention becomes kind of uniform and every token gets some form of poolled embeddings from others (with processing by k matrix) versus usual diagonal attention matrix. In the second version, the order of information per image is preserved but incorrectly reshaped. As a result, the feature embedding block learns to adapt to these perturbations and provides useful information (as accuracy drops if reshaped correctly those weights). 
 
Tensorboards clearly show that training curves are getting better for every improvement (including the one to come), we see up to 1% improvement at starting/intermediate epochs. At the last epoch the difference gets smaller. This hints that GA is really useful. Finally, there are more contributions than just GA: downsampling (with cross channel interaction), reparametrization (embedding, mlp ratio etc), in-stem design. All of those make the model better.

**Comment by rwightman** [comment]
With image size scaling, yes, vit is not great. But many of the architecture that would benefit from the improve resolution scaling (ie swin, this one) don't push much the standard resoultions though as they are still expensive to train well. Not trying to discredit the arch or ideas here, I do like the direction.

I have seen a lot of architectures, gone through them in detail, and as the 'mistake' here shows, sometimes it is less about the theoretical grounding behind an idea, and seemingly just adding more FLOPS or activations in the network with a good degree of mixing that matters.

Having seen a lot of arch, w/ data on their peformance, I can say these models do come out where you'd expect (in the upper half, above average, yes, but not breaking away from the pack).  You focus on the FLOPS and Param Counts in the paper, which are standard metrics.  But if we look at activation counts (for which they are relatively high, like swin), the performance is where you'd expect. And activation counts are very closely associated with GPU runtime performance... see table below picking a few points around the 30M activation count range (all 1k pretrain w/ no distill). Picked the points below from a graph that just had the macts + top-1, the img/sec was pulled up later and was remarkably consistent...

| model | MActs | top-1 | img/sec |
| --- | --- | --- | --- | 
| regnety_080 | 26.7 | 83.9 | 934 | 
| efficientnetv2_s | 35.8 | 83.9 | 1120 |
| deit3_base_patch16_224 | 24 | 83.8 | 1100 | 
| convnext_base |  28.8 | 83.8 | 1010 | 
| cs3se_edgnet_x | 20.1 | 83.5 | 1120 | 
| gcvit_tiny | 29.8 | 83.4 | 1010 | 
| swin_small_patch4_window7_224 | 27.5 | 83.2 | 1180 |

**Comment by ahatamiz** [comment]
Hi @rwightman , @leondgarse , @awsaf49 , @shkarupa-alex 

We would like to sincerely thank you for providing feedback, which has helped us to improve our work. We have updated the model and checkpoints (link ins ReadMe) which consistently demonstrate improved performance. Here's a summary:

| model | top-1 | top-5 |
| --- | --- | --- | 
| ViT-XXT | 79.80 | 95.09 |
| ViT-XT | 82.04 | 95.99 |
| ViT-T | 83.40 | 96.40 |
| ViT-S | 83.95 | 96.65 |
| ViT-B | 84.47 | 96.84 |


Best,

**Comment by awsaf49** [comment]
@ahatamiz I can clearly see some improvements (`~1%`) with new weights ....

**Comment by ahatamiz** [comment]
@awsaf49 Thanks for the comment. That's great news. I also tested it on different benchmarks and saw improvement across all experiments. 

Best,
